src/librustc_unicode/char.rs

```diff
@@ -234,22 +299,63 @@ impl char {
     #[inline]
     pub fn escape_default(self) -> EscapeDefault { C::escape_default(self) }
 
-    /// Returns the number of bytes this character would need if encoded in
-    /// UTF-8.
+    /// Returns the number of bytes the character this `char` would need if encoded in UTF-8.
+    ///
+    /// The `char` type is always four bytes, but not every character needs the full four when
+    /// encoded as UTF-8. `len_utf8` returns this length.
     ///
     /// # Examples
     ///
+    /// Basic usage:
+    ///
+    /// ```
+    /// let len = 'A'.len_utf8();
+    /// assert_eq!(len, 1);
+    ///
+    /// let len = 'ß'.len_utf8();
+    /// assert_eq!(len, 2);
+    ///
+    /// let len = 'ℝ'.len_utf8();
+    /// assert_eq!(len, 3);
+    ///
+    /// let len = '💣'.len_utf8();
+    /// assert_eq!(len, 4);
+    /// ```
+    ///
+    /// The `&str` type guarantees that its contents are UTF-8, and so we can compare the length it
+    /// would take if each codepoint was represented as a `char` vs in the `&str` itself:
+    ///
     /// ```
-    /// let n = 'ß'.len_utf8();
+    /// // as chars
+    /// let eastern = '東';
+    /// let capitol = '京';
+    ///
+    /// // both can be represented as three bytes...
+    /// assert_eq!(3, eastern.len_utf8());
+    /// assert_eq!(3, capitol.len_utf8());
+    ///
+    /// // ... but, as two chars, take eight bytes
+    /// assert_eq!(8, std::mem::size_of::<char>() * 2);
+    ///
+    /// // as a &str, these two are encoded in UTF-8
+    /// let tokyo = "東京";
+    ///
+    /// let len = eastern.len_utf8() + capitol.len_utf8();
     ///
-    /// assert_eq!(n, 2);
+    /// // we can see that they take six bytes total...
+    /// assert_eq!(6, tokyo.len());
+    ///
+    /// // ... just like the &str
+    /// assert_eq!(len, tokyo.len());
     /// ```
     #[stable(feature = "rust1", since = "1.0.0")]
     #[inline]
     pub fn len_utf8(self) -> usize { C::len_utf8(self) }
 
-    /// Returns the number of 16-bit code units this character would need if
-    /// encoded in UTF-16.
+    /// Returns the number of bytes the character this `char` would need if encoded in UTF-16.
```

Comment:
This is a number of 16-bit code units, not a number of bytes.

(Another difference from `len_utf8` is the the return value can only be 1 or 2, never 3 or 4.)